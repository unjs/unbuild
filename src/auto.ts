import { existsSync } from "node:fs";
import { normalize, join, resolve } from "pathe";
import { consola } from "consola";
import chalk from "chalk";
import type { PackageJson } from "pkg-types";
import { extractExportFilenames, listRecursively, warn } from "./utils";
import { BuildEntry, definePreset, MkdistBuildEntry } from "./types";

type InferEntriesResult = {
  entries: BuildEntry[];
  cjs?: boolean;
  dts?: boolean;
  warnings: string[];
};

export const autoPreset = definePreset(() => {
  return {
    hooks: {
      "build:prepare"(ctx) {
        // Disable auto if entries already provided or pkg not available
        if (!ctx.pkg || ctx.options.entries.length > 0) {
          return;
        }
        const sourceFiles = listRecursively(join(ctx.options.rootDir, "src"));
        const res = inferEntries(ctx.pkg, sourceFiles, ctx.options.rootDir);
        for (const message of res.warnings) {
          warn(ctx, message);
        }
        ctx.options.entries.push(...res.entries);
        if (res.cjs) {
          ctx.options.rollup.emitCJS = true;
        }
        if (res.dts) {
          ctx.options.declaration = res.dts;
        }
        consola.info(
          "Automatically detected entries:",
          chalk.cyan(
            ctx.options.entries
              .map((e) =>
                chalk.bold(
                  e.input
                    .replace(ctx.options.rootDir + "/", "")
                    .replace(/\/$/, "/*"),
                ),
              )
              .join(", "),
          ),
          chalk.gray(
            ["esm", res.cjs && "cjs", res.dts && "dts"]
              .filter(Boolean)
              .map((tag) => `[${tag}]`)
              .join(" "),
          ),
        );
      },
    },
  };
});

/**
 * @param {PackageJson} pkg The contents of a package.json file to serve as the source for inferred entries.
 * @param {string[]} sourceFiles A list of source files to use for inferring entries.
 * @param {string | undefined} rootDir The root directory of the project.
 */
export function inferEntries(
  pkg: PackageJson,
  sourceFiles: string[],
  rootDir?: string,
): InferEntriesResult {
  const warnings = [];

  // Sort files so least-nested files are first
  sourceFiles.sort((a, b) => a.split("/").length - b.split("/").length);

  // Come up with a list of all output files & their formats
  const outputs = extractExportFilenames(pkg.exports);

  if (pkg.bin) {
    const binaries =
      typeof pkg.bin === "string" ? [pkg.bin] : Object.values(pkg.bin);
    for (const file of binaries) {
      outputs.push({ file });
    }
  }
  if (pkg.main) {
    outputs.push({ file: pkg.main });
  }
  if (pkg.module) {
    outputs.push({ type: "esm", file: pkg.module });
  }
  if (pkg.types || pkg.typings) {
    outputs.push({ file: pkg.types || pkg.typings! });
  }

  // Try to detect output types
  const isESMPkg = pkg.type === "module";
  for (const output of outputs.filter((o) => !o.type)) {
    const isJS = output.file.endsWith(".js");
    if ((isESMPkg && isJS) || output.file.endsWith(".mjs")) {
      output.type = "esm";
    } else if ((!isESMPkg && isJS) || output.file.endsWith(".cjs")) {
      output.type = "cjs";
    }
  }

  let cjs = false;
  let dts = false;

  // Infer entries from package files
  const entries: BuildEntry[] = [];
  for (const output of outputs) {
    // Supported output file extensions are `.d.ts`, `.cjs` and `.mjs`
    // But we support any file extension here in case user has extended rollup options
    const outputSlug = output.file.replace(
      /(\*[^/\\]*|\.d\.(m|c)?ts|\.\w+)$/,
      "",
    );
    const isDir = outputSlug.endsWith("/");

    // Skip top level directory
    if (isDir && ["./", "/"].includes(outputSlug)) {
      continue;
    }

    const possiblePaths = getEntrypointPaths(outputSlug);
    // eslint-disable-next-line unicorn/no-array-reduce
    const input = possiblePaths.reduce<string | undefined>((source, d) => {
      if (source) {
        return source;
      }
      const SOURCE_RE = new RegExp(
        `(?<=/|$)${d}${isDir ? "" : String.raw`\.\w+`}$`,
      );
      return sourceFiles
        .find((i) => SOURCE_RE.test(i))
        ?.replace(/(\.d\.(m|c)?ts|\.\w+)$/, "");
    }, undefined as any);

    if (!input) {
      if (!existsSync(resolve(rootDir || ".", output.file))) {
        warnings.push(`Could not find entrypoint for \`${output.file}\``);
      }
      continue;
    }

    if (output.type === "cjs") {
      cjs = true;
    }

    const entry =
      entries.find((i) => i.input === input) ||
      entries[entries.push({ input }) - 1];

    if (/\.d\.(m|c)?ts$/.test(output.file)) {
      dts = true;
    }

    if (isDir) {
      entry.outDir = outputSlug;
      (entry as MkdistBuildEntry).format = output.type;
    }
  }

  return { entries, cjs, dts, warnings };
}

export const getEntrypointPaths = (path: string) => {
  const segments = normalize(path).split("/");
  return segments
    .map((_, index) => segments.slice(index).join("/"))
    .filter(Boolean);
};
