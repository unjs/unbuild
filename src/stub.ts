import jiti from 'jiti'

export { jiti }
