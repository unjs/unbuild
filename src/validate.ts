import { existsSync } from "node:fs";
import chalk from "chalk";
import { resolve } from "pathe";
import { PackageJson } from "pkg-types";
import { arrayIncludes, extractExportFilenames, getpkg, warn } from "./utils";
import { BuildContext } from "./types";

export function validateDependencies(ctx: BuildContext) {
  const usedDependencies = new Set<string>();
  const unusedDependencies = new Set<string>(
    Object.keys(ctx.pkg.dependencies || {})
  );
  const implicitDependencies = new Set<string>();
  for (const id of ctx.usedImports) {
    unusedDependencies.delete(id);
    usedDependencies.add(id);
  }
  if (Array.isArray(ctx.options.dependencies)) {
    for (const id of ctx.options.dependencies) {
      unusedDependencies.delete(id);
    }
  }
  for (const id of usedDependencies) {
    if (
      !arrayIncludes(ctx.options.externals, id) &&
      !id.startsWith("chunks/") &&
      !ctx.options.dependencies.includes(getpkg(id)) &&
      !ctx.options.peerDependencies.includes(getpkg(id))
    ) {
      implicitDependencies.add(id);
    }
  }
  if (unusedDependencies.size > 0) {
    warn(
      ctx,
      "Potential unused dependencies found: " +
        [...unusedDependencies].map((id) => chalk.cyan(id)).join(", ")
    );
  }
  if (implicitDependencies.size > 0 && !ctx.options.rollup.inlineDependencies) {
    warn(
      ctx,
      "Potential implicit dependencies found: " +
        [...implicitDependencies].map((id) => chalk.cyan(id)).join(", ")
    );
  }
}

export function validatePackage(
  pkg: PackageJson,
  rootDir: string,
  ctx: BuildContext
) {
  if (!pkg) {
    return;
  }

  const filenames = new Set(
    [
      ...(typeof pkg.bin === "string"
        ? [pkg.bin]
        : Object.values(pkg.bin || {})),
      pkg.main,
      pkg.module,
      pkg.types,
      pkg.typings,
      ...extractExportFilenames(pkg.exports).map((i) => i.file),
    ].map((i) => i && resolve(rootDir, i.replace(/\/[^/]*\*.*$/, "")))
  );

  const missingOutputs = [];

  for (const filename of filenames) {
    if (filename && !filename.includes("*") && !existsSync(filename)) {
      missingOutputs.push(filename.replace(rootDir + "/", ""));
    }
  }
  if (missingOutputs.length > 0) {
    warn(
      ctx,
      `Potential missing package.json files: ${missingOutputs
        .map((o) => chalk.cyan(o))
        .join(", ")}`
    );
  }
}
