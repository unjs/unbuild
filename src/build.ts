import Module from "node:module";
import { promises as fsp } from "node:fs";
import { resolve, relative, basename } from "pathe";
import type { PackageJson } from "pkg-types";
import chalk from "chalk";
import consola from "consola";
import { defu } from "defu";
import { createHooks } from "hookable";
import prettyBytes from "pretty-bytes";
import { globby } from "globby";
import mkdirp from "mkdirp";
import { dumpObject, rmdir, tryRequire, resolvePreset } from "./utils";
import type { BuildContext, BuildConfig, BuildOptions } from "./types";
import { validatePackage, validateDependencies } from "./validate";
import { rollupBuild } from "./builder/rollup";
import { typesBuild } from "./builder/untyped";
import { mkdistBuild } from "./builder/mkdist";

export async function build(
  rootDir: string,
  stub: boolean,
  inputConfig: BuildConfig = {}
) {
  // Determine rootDir
  rootDir = resolve(process.cwd(), rootDir || ".");

  // Read build.config and package.json
  const buildConfig: BuildConfig = tryRequire("./build.config", rootDir) || {};
  const pkg: PackageJson & Record<"unbuild" | "build", BuildConfig> =
    tryRequire("./package.json", rootDir);

  // Resolve preset
  const preset = resolvePreset(
    buildConfig.preset ||
      pkg.unbuild?.preset ||
      pkg.build?.preset ||
      inputConfig.preset ||
      "auto",
    rootDir
  );

  // Merge options
  const options = defu(
    buildConfig,
    pkg.unbuild || pkg.build,
    inputConfig,
    preset,
    <BuildOptions>{
      name: (pkg?.name || "").split("/").pop() || "default",
      rootDir,
      entries: [],
      clean: true,
      declaration: false,
      outDir: "dist",
      stub,
      externals: [
        ...Module.builtinModules,
        ...Module.builtinModules.map((m) => "node:" + m),
      ],
      dependencies: [],
      devDependencies: [],
      peerDependencies: [],
      alias: {},
      replace: {},
      failOnWarn: true,
      rollup: {
        emitCJS: false,
        cjsBridge: false,
        inlineDependencies: false,
        // Plugins
        replace: {
          preventAssignment: true,
        },
        alias: {},
        resolve: {
          preferBuiltins: true,
        },
        json: {
          preferConst: true,
        },
        commonjs: {
          ignoreTryCatch: true,
        },
        esbuild: { target: "es2020" },
        dts: {
          // https://github.com/Swatinem/rollup-plugin-dts/issues/143
          compilerOptions: { preserveSymlinks: false },
          respectExternal: true,
        },
      },
    }
  ) as BuildOptions;

  // Resolve dirs relative to rootDir
  options.outDir = resolve(options.rootDir, options.outDir);

  // Build context
  const ctx: BuildContext = {
    options,
    warnings: new Set(),
    pkg,
    buildEntries: [],
    usedImports: new Set(),
    hooks: createHooks(),
  };

  // Register hooks
  if (preset.hooks) {
    ctx.hooks.addHooks(preset.hooks);
  }
  if (inputConfig.hooks) {
    ctx.hooks.addHooks(inputConfig.hooks);
  }
  if (buildConfig.hooks) {
    ctx.hooks.addHooks(buildConfig.hooks);
  }

  // Allow prepare and extending context
  await ctx.hooks.callHook("build:prepare", ctx);

  // Normalize entries
  options.entries = options.entries.map((entry) =>
    typeof entry === "string" ? { input: entry } : entry
  );

  for (const entry of options.entries) {
    if (typeof entry.name !== "string") {
      entry.name = basename(entry.input);
    }

    if (!entry.input) {
      throw new Error("Missing entry input: " + dumpObject(entry));
    }

    if (!entry.builder) {
      entry.builder = entry.input.endsWith("/") ? "mkdist" : "rollup";
    }

    if (options.declaration !== undefined && entry.declaration === undefined) {
      entry.declaration = options.declaration;
    }

    entry.input = resolve(options.rootDir, entry.input);
    entry.outDir = resolve(options.rootDir, entry.outDir || options.outDir);
  }

  // Infer dependencies from pkg
  options.dependencies = Object.keys(pkg.dependencies || {});
  options.peerDependencies = Object.keys(pkg.peerDependencies || {});
  options.devDependencies = Object.keys(pkg.devDependencies || {});

  // Add all dependencies as externals
  options.externals.push(...options.dependencies, ...options.peerDependencies);

  // Call build:before
  await ctx.hooks.callHook("build:before", ctx);

  // Start info
  consola.info(
    chalk.cyan(`${options.stub ? "Stubbing" : "Building"} ${pkg.name}`)
  );
  if (process.env.DEBUG) {
    consola.info(`${chalk.bold("Root dir:")} ${options.rootDir}
  ${chalk.bold("Entries:")}
  ${options.entries.map((entry) => "  " + dumpObject(entry)).join("\n  ")}
`);
  }

  // Clean dist dirs
  if (options.clean) {
    for (const dir of new Set(options.entries.map((e) => e.outDir).sort())) {
      await rmdir(dir!);
      await mkdirp(dir!);
    }
  }

  // Try to selflink
  // if (ctx.stub && ctx.pkg.name) {
  //   const nodemodulesDir = resolve(ctx.rootDir, 'node_modules', ctx.pkg.name)
  //   await symlink(resolve(ctx.rootDir), nodemodulesDir).catch(() => {})
  // }

  // untyped
  await typesBuild(ctx);

  // mkdist
  await mkdistBuild(ctx);

  // rollup
  await rollupBuild(ctx);

  // Skip rest for stub
  if (options.stub) {
    await ctx.hooks.callHook("build:done", ctx);
    return;
  }

  // Done info
  consola.success(chalk.green("Build succeeded for " + options.name));

  // Find all dist files and add missing entries as chunks
  const outFiles = await globby("**", { cwd: options.outDir });
  for (const file of outFiles) {
    let entry = ctx.buildEntries.find((e) => e.path === file);
    if (!entry) {
      entry = {
        path: file,
        chunk: true,
      };
      ctx.buildEntries.push(entry);
    }
    if (!entry.bytes) {
      const stat = await fsp.stat(resolve(options.outDir, file));
      entry.bytes = stat.size;
    }
  }

  const rPath = (p: string) =>
    relative(process.cwd(), resolve(options.outDir, p));
  for (const entry of ctx.buildEntries.filter((e) => !e.chunk)) {
    let totalBytes = entry.bytes || 0;
    for (const chunk of entry.chunks || []) {
      totalBytes += ctx.buildEntries.find((e) => e.path === chunk)?.bytes || 0;
    }
    let line =
      `  ${chalk.bold(rPath(entry.path))} (` +
      [
        entry.bytes && `size: ${chalk.cyan(prettyBytes(entry.bytes))}`,
        totalBytes !== entry.bytes &&
          `total size: ${chalk.cyan(prettyBytes(totalBytes))}`,
        entry.exports?.length &&
          `exports: ${chalk.gray(entry.exports.join(", "))}`,
      ]
        .filter(Boolean)
        .join(", ") +
      ")";
    if (entry.chunks?.length) {
      line +=
        "\n" +
        entry.chunks
          .map((p) => {
            const chunk =
              ctx.buildEntries.find((e) => e.path === p) || ({} as any);
            return chalk.gray(
              "  └─ " +
                rPath(p) +
                (chunk.bytes ? ` (${prettyBytes(chunk?.bytes)})` : "")
            );
          })
          .join("\n");
    }
    consola.log(entry.chunk ? chalk.gray(line) : line);
  }
  console.log(
    "Σ Total dist size (byte size):",
    chalk.cyan(
      prettyBytes(ctx.buildEntries.reduce((a, e) => a + (e.bytes || 0), 0))
    )
  );

  // Validate
  validateDependencies(ctx);
  validatePackage(pkg, rootDir, ctx);

  // Call build:done
  await ctx.hooks.callHook("build:done", ctx);

  consola.log("");

  if (ctx.warnings.size > 0) {
    consola.warn(
      "Build is done with some warnings:\n\n" +
        [...ctx.warnings].map((msg) => "- " + msg).join("\n")
    );
    if (ctx.options.failOnWarn) {
      consola.error(
        "Exiting with code (1). You can change this behavior by setting `failOnWarn: false` ."
      );
      // eslint-disable-next-line unicorn/no-process-exit
      process.exit(1);
    }
  }
}
