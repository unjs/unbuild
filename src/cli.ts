#!/usr/bin/env node
import { resolve } from "pathe";
import mri from "mri";
import { build } from "./build";

async function main() {
  const args = mri(process.argv.splice(2));
  const rootDir = resolve(process.cwd(), args._[0] || ".");
  await build(rootDir, args.stub).catch((error) => {
    console.error(`Error building ${rootDir}: ${error}`);
    throw error;
  });
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
