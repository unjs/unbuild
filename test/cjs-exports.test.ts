import { describe, it, expect } from "vitest";
import { build } from "../src";
import { resolve } from "pathe";
import { fileURLToPath } from "node:url";
import { readdir, readFile } from "node:fs/promises";
import {
  type ESMExport,
  findExports,
  findStaticImports,
  findTypeExports,
  parseStaticImport,
  type StaticImport,
} from "mlly";

describe("Node10 and Node16 Default Exports Types", () => {
  const dtsFiles = /\.d\.(c)?ts$/;

  async function readDtsFiles(
    dist: string,
  ): Promise<
    [
      name: string,
      types: ESMExport[],
      exports: ESMExport[],
      content: string,
      imports: StaticImport[],
    ][]
  > {
    const files = await readdir(dist).then((files) =>
      files.filter((f) => dtsFiles.test(f)).map((f) => [f, resolve(dist, f)]),
    );
    return await Promise.all(
      files.map(async ([name, path]) => {
        const content = await readFile(path, "utf8");
        return [
          name,
          findTypeExports(content),
          findExports(content),
          content,
          findStaticImports(content),
        ];
      }),
    );
  }

  it("Mixed Declaration Types", async () => {
    const root = resolve(
      fileURLToPath(import.meta.url),
      "../cjs-types-fixture/mixed-declarations",
    );
    await build(root, false);
    const files = await readDtsFiles(resolve(root, "dist"));
    expect(files).toHaveLength(2);
    for (const [name, types, exports] of files) {
      expect(exports).toHaveLength(0);
      expect(types).toHaveLength(1);
      expect(
        types.find((e) => e.names.includes("default")),
        `${name} should not have a default export`,
      ).toBeUndefined();
    }
    expect(files).toMatchSnapshot();
  });

  it("Re-Export Types", async () => {
    const warnings: string[] = [];
    const root = resolve(
      fileURLToPath(import.meta.url),
      "../cjs-types-fixture/reexport-types",
    );
    await build(root, false, {
      hooks: {
        "rollup:options": (_, options) => {
          const _onwarn = options.onwarn;
          // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
          options.onwarn = (warning, handler) => {
            if (warning.code === "EMPTY_BUNDLE") {
              warnings.push(warning.message);
              return;
            }
            return _onwarn?.(warning, handler);
          };
        },
      },
    });
    expect(warnings).toHaveLength(2);
    expect(warnings[0]).toBe('Generated an empty chunk: "types".');
    expect(warnings[1]).toBe('Generated an empty chunk: "types".');
    const files = await readDtsFiles(resolve(root, "dist"));
    expect(files).toHaveLength(6);
    for (const [name, types, exports, content, imports] of files) {
      if (name.startsWith("types")) {
        expect(exports).toHaveLength(0);
        expect(types).toHaveLength(1);
        expect(
          types.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
      } else if (name.startsWith("index")) {
        expect(exports).toHaveLength(2);
        expect(types).toHaveLength(0);
        expect(imports).toHaveLength(1);
        expect(
          exports.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        expect(content).toMatch("export = plugin");
      } else if (name.startsWith("all")) {
        expect(exports).toHaveLength(2);
        expect(types).toHaveLength(0);
        expect(imports).toHaveLength(1);
        expect(
          exports.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        const defaultImport = parseStaticImport(imports[0]);
        expect(defaultImport.defaultImport).toBe("_default");
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`);
      }
    }
    expect(files).toMatchSnapshot();
  });

  it("Re-Export as default", async () => {
    const root = resolve(
      fileURLToPath(import.meta.url),
      "../cjs-types-fixture/reexport-default",
    );
    await build(root, false);
    const files = await readDtsFiles(resolve(root, "dist"));
    expect(files).toHaveLength(8);
    for (const [name, types, exports, content, imports] of files) {
      if (name.startsWith("asdefault")) {
        expect(exports).toHaveLength(0);
        expect(types).toHaveLength(0);
        expect(imports).toHaveLength(1);
        expect(
          types.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        const defaultImport = parseStaticImport(imports[0]);
        expect(defaultImport.namedImports?.resolve).toBeDefined();
        expect(content).toMatch(`export = resolve`);
      } else if (name.startsWith("index")) {
        expect(exports).toHaveLength(1);
        expect(types).toHaveLength(0);
        expect(imports).toHaveLength(1);
        expect(
          exports.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        const defaultImport = parseStaticImport(imports[0]);
        expect(defaultImport.defaultImport).toBe("MagicString");
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`);
      } else if (name.startsWith("magicstringasdefault")) {
        expect(exports).toHaveLength(0);
        expect(types).toHaveLength(0);
        expect(imports.filter((i) => !!i.imports)).toHaveLength(1);
        expect(
          exports.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        const defaultImport = parseStaticImport(imports[0]);
        expect(defaultImport.defaultImport).toBe("_default");
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`);
      } else if (name.startsWith("resolvedasdefault")) {
        expect(exports).toHaveLength(0);
        expect(types).toHaveLength(0);
        expect(imports.filter((i) => !!i.imports)).toHaveLength(1);
        expect(
          exports.find((e) => e.names.includes("default")),
          `${name} should not have a default export`,
        ).toBeUndefined();
        const defaultImport = parseStaticImport(imports[0]);
        expect(defaultImport.defaultImport).toBe("resolve");
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`);
      }
    }
    expect(files).toMatchSnapshot();
  });
});
