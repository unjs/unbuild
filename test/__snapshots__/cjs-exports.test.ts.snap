// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Node10 and Node16 Default Exports Types > Mixed Declaration Types 1`] = `
[
  [
    "index.d.cts",
    [
      {
        "code": "export type { A, B, C, Options }",
        "end": 260,
        "exports": " A, B, C, Options",
        "names": [
          "A",
          "B",
          "C",
          "Options",
        ],
        "specifier": undefined,
        "start": 228,
        "type": "named",
      },
    ],
    [],
    "interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}
interface Options {
    a?: A;
    b?: B;
    c?: C;
}
declare function plugin(options?: Options): Options;

export = plugin;
export type { A, B, C, Options };
",
    [],
  ],
  [
    "index.d.ts",
    [
      {
        "code": "export type { A, B, C, Options }",
        "end": 260,
        "exports": " A, B, C, Options",
        "names": [
          "A",
          "B",
          "C",
          "Options",
        ],
        "specifier": undefined,
        "start": 228,
        "type": "named",
      },
    ],
    [],
    "interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}
interface Options {
    a?: A;
    b?: B;
    c?: C;
}
declare function plugin(options?: Options): Options;

export = plugin;
export type { A, B, C, Options };
",
    [],
  ],
]
`;

exports[`Node10 and Node16 Default Exports Types > Re-Export Types 1`] = `
[
  [
    "all.d.cts",
    [],
    [
      {
        "code": "export { ResolvedOptions, default as plugin } from './index.cjs'",
        "end": 119,
        "exports": " ResolvedOptions, default as plugin",
        "names": [
          "ResolvedOptions",
          "plugin",
        ],
        "specifier": "./index.cjs",
        "start": 55,
        "type": "named",
      },
      {
        "code": "export { A, B, CC as C, CC, Options } from './types.cjs'",
        "end": 177,
        "exports": " A, B, CC as C, CC, Options",
        "names": [
          "A",
          "B",
          "C",
          "CC",
          "Options",
        ],
        "specifier": "./types.cjs",
        "start": 121,
        "type": "named",
      },
    ],
    "import _default from './index.cjs';
export = _default;
export { ResolvedOptions, default as plugin } from './index.cjs';
export { A, B, CC as C, CC, Options } from './types.cjs';
",
    [
      {
        "code": "import _default from './index.cjs';
",
        "end": 36,
        "imports": "_default ",
        "specifier": "./index.cjs",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "all.d.ts",
    [],
    [
      {
        "code": "export { ResolvedOptions, default as plugin } from './index.js'",
        "end": 117,
        "exports": " ResolvedOptions, default as plugin",
        "names": [
          "ResolvedOptions",
          "plugin",
        ],
        "specifier": "./index.js",
        "start": 54,
        "type": "named",
      },
      {
        "code": "export { A, B, CC as C, CC, Options } from './types.js'",
        "end": 174,
        "exports": " A, B, CC as C, CC, Options",
        "names": [
          "A",
          "B",
          "C",
          "CC",
          "Options",
        ],
        "specifier": "./types.js",
        "start": 119,
        "type": "named",
      },
    ],
    "import _default from './index.js';
export = _default;
export { ResolvedOptions, default as plugin } from './index.js';
export { A, B, CC as C, CC, Options } from './types.js';
",
    [
      {
        "code": "import _default from './index.js';
",
        "end": 35,
        "imports": "_default ",
        "specifier": "./index.js",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "index.d.cts",
    [],
    [
      {
        "code": "export { A, B, CC as C, CC } from './types.cjs'",
        "end": 86,
        "exports": " A, B, CC as C, CC",
        "names": [
          "A",
          "B",
          "C",
          "CC",
        ],
        "specifier": "./types.cjs",
        "start": 39,
        "type": "named",
      },
      {
        "code": "export { Options, type ResolvedOptions, plugin }",
        "end": 280,
        "exports": " Options, type ResolvedOptions, plugin",
        "names": [
          "Options",
          "plugin",
        ],
        "specifier": undefined,
        "start": 232,
        "type": "named",
      },
    ],
    "import { Options } from './types.cjs';
export { A, B, CC as C, CC } from './types.cjs';

interface ResolvedOptions extends Options {
    name: string;
}
declare function plugin(options?: Options): ResolvedOptions;

export = plugin;
export { Options, type ResolvedOptions, plugin };
",
    [
      {
        "code": "import { Options } from './types.cjs';
",
        "end": 39,
        "imports": "{ Options } ",
        "specifier": "./types.cjs",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "index.d.ts",
    [],
    [
      {
        "code": "export { A, B, CC as C, CC } from './types.js'",
        "end": 84,
        "exports": " A, B, CC as C, CC",
        "names": [
          "A",
          "B",
          "C",
          "CC",
        ],
        "specifier": "./types.js",
        "start": 38,
        "type": "named",
      },
      {
        "code": "export { Options, type ResolvedOptions, plugin }",
        "end": 278,
        "exports": " Options, type ResolvedOptions, plugin",
        "names": [
          "Options",
          "plugin",
        ],
        "specifier": undefined,
        "start": 230,
        "type": "named",
      },
    ],
    "import { Options } from './types.js';
export { A, B, CC as C, CC } from './types.js';

interface ResolvedOptions extends Options {
    name: string;
}
declare function plugin(options?: Options): ResolvedOptions;

export = plugin;
export { Options, type ResolvedOptions, plugin };
",
    [
      {
        "code": "import { Options } from './types.js';
",
        "end": 38,
        "imports": "{ Options } ",
        "specifier": "./types.js",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "types.d.cts",
    [
      {
        "code": "export type { A, B, C, C as CC, Options }",
        "end": 200,
        "exports": " A, B, C, C as CC, Options",
        "names": [
          "A",
          "B",
          "C",
          "CC",
          "Options",
        ],
        "specifier": undefined,
        "start": 159,
        "type": "named",
      },
    ],
    [],
    "interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}

interface Options {
    a?: A;
    b?: B;
    c?: C;
}

export type { A, B, C, C as CC, Options };
",
    [],
  ],
  [
    "types.d.ts",
    [
      {
        "code": "export type { A, B, C, C as CC, Options }",
        "end": 200,
        "exports": " A, B, C, C as CC, Options",
        "names": [
          "A",
          "B",
          "C",
          "CC",
          "Options",
        ],
        "specifier": undefined,
        "start": 159,
        "type": "named",
      },
    ],
    [],
    "interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}

interface Options {
    a?: A;
    b?: B;
    c?: C;
}

export type { A, B, C, C as CC, Options };
",
    [],
  ],
]
`;

exports[`Node10 and Node16 Default Exports Types > Re-Export as default 1`] = `
[
  [
    "asdefault.d.cts",
    [],
    [],
    "import { resolve } from 'pathe';
export = resolve;
",
    [
      {
        "code": "import { resolve } from 'pathe';
",
        "end": 33,
        "imports": "{ resolve } ",
        "specifier": "pathe",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "asdefault.d.ts",
    [],
    [],
    "import { resolve } from 'pathe';
export = resolve;
",
    [
      {
        "code": "import { resolve } from 'pathe';
",
        "end": 33,
        "imports": "{ resolve } ",
        "specifier": "pathe",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "defaultclass.d.cts",
    [],
    [],
    "declare class DefaultClass {
    name: string;
    constructor(name: string);
}

export = DefaultClass;
",
    [],
  ],
  [
    "defaultclass.d.ts",
    [],
    [],
    "declare class DefaultClass {
    name: string;
    constructor(name: string);
}

export = DefaultClass;
",
    [],
  ],
  [
    "index.d.cts",
    [],
    [
      {
        "code": "export { MagicStringOptions } from 'magic-string'",
        "end": 111,
        "exports": " MagicStringOptions",
        "name": "MagicStringOptions",
        "names": [
          "MagicStringOptions",
        ],
        "specifier": "magic-string",
        "start": 62,
        "type": "named",
      },
    ],
    "import MagicString from 'magic-string';
export = MagicString;
export { MagicStringOptions } from 'magic-string';
",
    [
      {
        "code": "import MagicString from 'magic-string';
",
        "end": 40,
        "imports": "MagicString ",
        "specifier": "magic-string",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "index.d.ts",
    [],
    [
      {
        "code": "export { MagicStringOptions } from 'magic-string'",
        "end": 111,
        "exports": " MagicStringOptions",
        "name": "MagicStringOptions",
        "names": [
          "MagicStringOptions",
        ],
        "specifier": "magic-string",
        "start": 62,
        "type": "named",
      },
    ],
    "import MagicString from 'magic-string';
export = MagicString;
export { MagicStringOptions } from 'magic-string';
",
    [
      {
        "code": "import MagicString from 'magic-string';
",
        "end": 40,
        "imports": "MagicString ",
        "specifier": "magic-string",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "magicstringasdefault.d.cts",
    [],
    [],
    "import _default from 'magic-string';
export = _default;
import 'pathe';
",
    [
      {
        "code": "import _default from 'magic-string';
",
        "end": 37,
        "imports": "_default ",
        "specifier": "magic-string",
        "start": 0,
        "type": "static",
      },
      {
        "code": "import 'pathe';
",
        "end": 72,
        "imports": undefined,
        "specifier": "pathe",
        "start": 56,
        "type": "static",
      },
    ],
  ],
  [
    "magicstringasdefault.d.ts",
    [],
    [],
    "import _default from 'magic-string';
export = _default;
import 'pathe';
",
    [
      {
        "code": "import _default from 'magic-string';
",
        "end": 37,
        "imports": "_default ",
        "specifier": "magic-string",
        "start": 0,
        "type": "static",
      },
      {
        "code": "import 'pathe';
",
        "end": 72,
        "imports": undefined,
        "specifier": "pathe",
        "start": 56,
        "type": "static",
      },
    ],
  ],
  [
    "resolveasdefault.d.cts",
    [],
    [],
    "import { resolve } from 'pathe';
export = resolve;
",
    [
      {
        "code": "import { resolve } from 'pathe';
",
        "end": 33,
        "imports": "{ resolve } ",
        "specifier": "pathe",
        "start": 0,
        "type": "static",
      },
    ],
  ],
  [
    "resolveasdefault.d.ts",
    [],
    [],
    "import { resolve } from 'pathe';
export = resolve;
",
    [
      {
        "code": "import { resolve } from 'pathe';
",
        "end": 33,
        "imports": "{ resolve } ",
        "specifier": "pathe",
        "start": 0,
        "type": "static",
      },
    ],
  ],
]
`;
